pandas/core/indexes/base.py

```diff
@@ -2939,7 +2964,11 @@ def union(self, other, sort=None):
                 # float | [u]int -> float  (the special case)
                 # <T>   | <T>    -> T
                 # <T>   | <U>    -> object
-                if not (is_integer_dtype(self.dtype) and is_integer_dtype(other.dtype)):
+                if is_float_dtype(self.dtype) and is_float_dtype(other.dtype):
+                    dtype = np.dtype("float64")
```

Comment:
`find_common_type(int32, float63)` would return float64, the comment above indicates that this currently is treated special and be an object dtype..

If the current behaviour is a error, I think it would probably be better to fix that separately.

Reply:
> comment above indicates that this currently is treated special and be an object dtype..

the special case is uint64 vs int64, where if we used the find_common_dtype (float64) it would be potentially lossy.  int32 vs float64 the conversion to float64 should always be lossless

Reply:
The return value from `find_common_type([np.int64, np.uint64])` is a float64. Would that be more wrong than a an object type?

Also, `find_common_type((np.int32, np.uint64))` returna float64 also, is that ok to use?

Reply:
> The return value from find_common_type([np.int64, np.uint64]) is a float64. Would that be more wrong than a an object type?

im not clear on what you're asking.  are you asking if find_common_type is wrong?

> Also, find_common_type((np.int32, np.uint64)) returna float64 also, is that ok to use?

for it to be non-lossy that'll have to go to object

Reply:
Ok, I think I understand your question. You're asking if line 2967 can be replaced by `if find_common_type([self.dtype, other.dtype]) == np.float64:` or similar?

I think that's not possible without accepting lossyness, because `find_common_type` of ints resolves to float64 in many cases. But many series int operations do resolve to float64, so pandas does accept lossiness. OTOH, lossiness in probably not a good idea in union operations, so lossiness should be avoided here, don't you agree?

Reply:
stepping back, L2934-L2946 (in master, not PR) are best thought of as

```
dtype = find_common_type([self.dtype, other.dtype])
if not is_lossless_casting([self.dtype, other.dtype], dtype):
    # comment about why we treat this differently than find_common_type
    dtype = np.dtype("object")
```

so pinning down the correct edit is equivalent to writing a full implementation of `is_lossless_casting`.

Are we on the same page?

Reply:
Ok, I think I've pinned this down now:

We want to in general to follow `find_common_type`. E.g. if self.dtype is a np.uint8 and other.dtype is a np.int8, we want np.int16. An exception is that we want a object dtype, if both are integers, but find_common_type does not resolve to an integer, because that must mean that one is an uint64 and the other is a signed int. for example:

* (np.int8, np.uint32) -> np.int64
* (np.int8, np.uint64) -> object

In the second example `find_common_type` would return a float64, which we would not want.

I've changed the implementation.

Reply:
Likewise fixed in #41773.